src/proto/message.rs

```diff
@@ -1,134 +1,339 @@
+        .checked_sum()
+    }
+
+    fn encode(&self, writer: &mut impl Writer) -> ssh_encoding::Result<()> {
+        self.pubkey.encode(writer)?;
+        self.data.encode(writer)?;
+        self.flags.encode(writer)?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct AddIdentity {
-    pub privkey: PrivateKey,
+    pub privkey: KeypairData,
     pub comment: String,
 }
 
-#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+impl Decode for AddIdentity {
+    type Error = Error;
+
+    fn decode(reader: &mut impl Reader) -> Result<Self> {
+        let privkey = KeypairData::decode(reader)?;
+        let comment = String::decode(reader)?;
+
+        Ok(Self { privkey, comment })
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct AddIdentityConstrained {
     pub identity: AddIdentity,
     pub constraints: Vec<KeyConstraint>,
 }
 
-#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+impl Decode for AddIdentityConstrained {
+    type Error = Error;
+
+    fn decode(reader: &mut impl Reader) -> Result<Self> {
+        let identity = AddIdentity::decode(reader)?;
+        let mut constraints = vec![];
+
+        while !reader.is_finished() {
+            constraints.push(KeyConstraint::decode(reader)?);
+        }
+        Ok(Self {
+            identity,
+            constraints,
+        })
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct RemoveIdentity {
-    pub pubkey_blob: Vec<u8>,
+    pub pubkey: KeyData,
+}
+
+impl Decode for RemoveIdentity {
+    type Error = Error;
+
+    fn decode(reader: &mut impl Reader) -> Result<Self> {
+        let pubkey = reader.read_prefixed(KeyData::decode)?;
+
+        Ok(Self { pubkey })
+    }
 }
 
-#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+#[derive(Clone, PartialEq, Debug)]
 pub struct SmartcardKey {
     pub id: String,
     pub pin: String,
 }
 
-#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
-pub struct KeyConstraint {
-    pub constraint_type: u8,
-    pub constraint_data: Vec<u8>,
+impl Decode for SmartcardKey {
+    type Error = Error;
+
+    fn decode(reader: &mut impl Reader) -> Result<Self> {
+        let id = String::decode(reader)?;
+        let pin = String::decode(reader)?;
+
+        Ok(Self { id, pin })
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum KeyConstraint {
+    Lifetime(u32),
+    Confirm,
+    Extension(String, Vec<u8>),
 }
 
-#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
+impl Decode for KeyConstraint {
+    type Error = Error;
+
+    fn decode(reader: &mut impl Reader) -> Result<Self> {
+        let constraint_type = u8::decode(reader)?;
+        // see: https://www.ietf.org/archive/id/draft-miller-ssh-agent-12.html#section-5.2
+        Ok(match constraint_type {
+            1 => KeyConstraint::Lifetime(u32::decode(reader)?),
+            2 => KeyConstraint::Confirm,
+            255 => KeyConstraint::Extension(String::decode(reader)?, Vec::<u8>::decode(reader)?),
```

Comment:
lovely <3
This makes it much more flexible! I didn't realized we had the size of the vec in the payload and that we didn't require a proper serialization implementation of the content.

Reply:
I actually didn't test this yet but I found ssh-add has a destination constraint option so I'll capture some bytes for unit tests tomorrow (just like I did for SessionBind now).

In general I'll adjust key constraints extensions to work like regular extensions (see code below with Unparsed). That is it should be possible to deserialize them even if you define them outside of this crate. This gives the consumers maximum flexibility.